OwningComponentBase implements IAsyncDisposable

rolandVi opened the pull request:
# OwningComponentBase implements IAsyncDisposable
## Description

This pull request enhances the `OwningComponentBase` class by adding support for asynchronous disposal, improving resource management, and updating tests to validate the new functionality. The most important changes include implementing the `IAsyncDisposable` interface, adding new methods for asynchronous disposal, updating the public API, and introducing new test cases to ensure proper behavior.

### Enhancements to `OwningComponentBase`:

* [`src/Components/Components/src/OwningComponentBase.cs`](diffhunk://#diff-82e944331c19e7b8ef290d4273e29da3e222ac5ed6df259a5d06de8dceaba6d0L17-R17): Updated the `OwningComponentBase` class to implement the `IAsyncDisposable` interface. Added `DisposeAsync`, `DisposeAsyncCore`, and updated the `Dispose` method to handle both synchronous and asynchronous disposal of the service scope.
* The changes were made based on this [guidance](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#implement-both-dispose-and-async-dispose-patterns).

### Public API updates:

* [`src/Components/Components/src/PublicAPI.Unshipped.txt`](diffhunk://#diff-5dde1dcc82e9e2c2b4f749af8c46b41f8977be2499ef6d62ef74fca5628b6e9dR3): Added `DisposeAsync` and `DisposeAsyncCore` methods to the public API to reflect the new asynchronous disposal capabilities.

### Testing improvements:

* [`src/Components/Components/test/OwningComponentBaseTest.cs`](diffhunk://#diff-964abc7e153a8de645f2be6ff4c49b8fb02d4e4cc9c65a1f4f91f6bde11ba398R32-R94): Added new test cases to verify the behavior of `DisposeAsync`, ensure that accessing scoped services after disposal throws an exception.

Fixes #25873

Commit:
Added async-disposable functionality to OwningComponentBase + tests

rolandVi committed:
Merge branch 'main' into OwningComponentBase-implements-IAsyncDisposable

ilonatommy commented on src/Components/Components/src/PublicAPI.Unshipped.txt:
```diff
@@ -19,5 +20,7 @@ Microsoft.AspNetCore.Components.Infrastructure.PersistentStateProviderServiceCol
 static Microsoft.AspNetCore.Components.Infrastructure.RegisterPersistentComponentStateServiceCollectionExtensions.AddPersistentServiceRegistration<TService>(Microsoft.Extensions.DependencyInjection.IServiceCollection! services, Microsoft.AspNetCore.Components.IComponentRenderMode! componentRenderMode) -> Microsoft.Extensions.DependencyInjection.IServiceCollection!
 static Microsoft.AspNetCore.Components.Infrastructure.ComponentsMetricsServiceCollectionExtensions.AddComponentsMetrics(Microsoft.Extensions.DependencyInjection.IServiceCollection! services) -> Microsoft.Extensions.DependencyInjection.IServiceCollection!
 static Microsoft.AspNetCore.Components.Infrastructure.ComponentsMetricsServiceCollectionExtensions.AddComponentsTracing(Microsoft.Extensions.DependencyInjection.IServiceCollection! services) -> Microsoft.Extensions.DependencyInjection.IServiceCollection!
+static Microsoft.Extensions.DependencyInjection.SupplyParameterFromPersistentComponentStateProviderServiceCollectionExtensions.AddSupplyValueFromPersistentComponentStateProvider(this Microsoft.Extensions.DependencyInjection.IServiceCollection! services) -> Microsoft.Extensions.DependencyInjection.IServiceCollection!
```
This was unintentional removal, let's revert it.

rolandVi committed:
Update PublicAPI.Unshipped.txt

Reverted one change

javiercn commented on src/Components/Components/src/OwningComponentBase.cs:
```diff
@@ -44,20 +44,63 @@ protected IServiceProvider ScopedServices
+    /// Asynchronously releases the service scope used by the component.
+    /// </summary>
+    /// <returns>A task that represents the asynchronous dispose operation.</returns>
+    public async ValueTask DisposeAsync()
```
Make the implementation private. Like the one for Dispose.

https://github.com/dotnet/aspnetcore/pull/62583/files#diff-82e944331c19e7b8ef290d4273e29da3e222ac5ed6df259a5d06de8dceaba6d0R50

To make an interface implementation private, you remove public and put the interface name right before the method `(in this case IAsyncDisposable.DisposeAsync)`

Commit:
DisposeAsync implementation made private + style restructure

rolandVi committed:
Update PublicAPI.Unshipped.txt

Removed `OwningComponentBase.DisposeAsync()` from the PublicAPI as the implementation is private now

javiercn reviewed (approved):
Looks great!

Commit:
Doc + style changes

rolandVi commented on src/Components/Components/src/OwningComponentBase.cs:
```diff
@@ -44,20 +44,63 @@ protected IServiceProvider ScopedServices
 
+    /// <summary>
+    /// Releases the service scope used by the component.
+    /// </summary>
```
Thanks for the explanation, I understand now. It's fixed now

rolandVi commented:
/ba-g failure connected with known issue
